diff --git old/lib/Driver/ToolChains/Ananas.cpp clang/lib/Driver/ToolChains/Ananas.cpp
index 006fdc029e..c779395e2f 100644
--- old/lib/Driver/ToolChains/Ananas.cpp
+++ clang/lib/Driver/ToolChains/Ananas.cpp
@@ -134,6 +134,23 @@ Ananas::Ananas(const Driver &D, const llvm::Triple &Triple, const ArgList &Args)
   getFilePaths().push_back(getDriver().SysRoot + "/usr/lib");
 }
 
+void Ananas::addLibCxxIncludePaths(const llvm::opt::ArgList &DriverArgs,
+                                     llvm::opt::ArgStringList &CC1Args) const {
+
+  StringRef SysRoot = getDriver().SysRoot;
+
+  SmallString<128> Dir(SysRoot);
+  llvm::sys::path::append(Dir, "usr", "include", "c++", "v1");
+  addSystemInclude(DriverArgs, CC1Args, Dir.str());
+}
+
+void Ananas::AddCXXStdlibLibArgs(const ArgList &Args,
+                                   ArgStringList &CmdArgs) const {
+  CmdArgs.push_back("-lc++");
+  CmdArgs.push_back("-lc++abi");
+  CmdArgs.push_back("-lunwind");
+}
+
 Tool *Ananas::buildAssembler() const {
   return new tools::ananas::Assembler(*this);
 }
diff --git old/lib/Driver/ToolChains/Ananas.h clang/lib/Driver/ToolChains/Ananas.h
index 2563dd2d49..b0c2133f93 100644
--- old/lib/Driver/ToolChains/Ananas.h
+++ clang/lib/Driver/ToolChains/Ananas.h
@@ -55,6 +55,16 @@ public:
   Ananas(const Driver &D, const llvm::Triple &Triple,
          const llvm::opt::ArgList &Args);
 
+  CXXStdlibType
+  GetCXXStdlibType(const llvm::opt::ArgList &Args) const override {
+    return ToolChain::CST_Libcxx;
+  }
+  void addLibCxxIncludePaths(
+      const llvm::opt::ArgList &DriverArgs,
+      llvm::opt::ArgStringList &CC1Args) const override;
+  void AddCXXStdlibLibArgs(const llvm::opt::ArgList &Args,
+                           llvm::opt::ArgStringList &CmdArgs) const override;
+
 protected:
   Tool *buildAssembler() const override;
   Tool *buildLinker() const override;
