diff -rubB coreutils-8.23.org/build-aux/config.sub coreutils-8.23/build-aux/config.sub
--- coreutils-8.23.org/build-aux/config.sub	2014-06-25 10:43:35.000000000 +0200
+++ coreutils-8.23/build-aux/config.sub	2015-04-27 10:08:13.304243419 +0200
@@ -1512,6 +1512,8 @@
 		;;
 	-nacl*)
 		;;
+	-ananas*)
+		;;
 	-none)
 		;;
 	*)
diff -rubB coreutils-8.23.org/lib/freading.c coreutils-8.23/lib/freading.c
--- coreutils-8.23.org/lib/freading.c	2014-05-29 14:05:50.000000000 +0200
+++ coreutils-8.23/lib/freading.c	2015-04-27 10:10:12.084247057 +0200
@@ -66,6 +66,8 @@
   if (fp->state == 0 /* CLOSED */ || fp->state == 4 /* WR */)
     return 0;
   return (fp->state == 3 /* RD */ && (fp->bufl == 0 || fp->rp < fp->wp));
+#elif defined _PDCLIB_C_VERSION     /* PDCLIBC */
+  return (fp->status & _PDCLIB_FREAD) != 0;
 # else
 #  error "Please port gnulib freading.c to your platform!"
 # endif
diff -rubB coreutils-8.23.org/lib/fseeko.c coreutils-8.23/lib/fseeko.c
--- coreutils-8.23.org/lib/fseeko.c	2014-07-18 20:31:54.000000000 +0200
+++ coreutils-8.23/lib/fseeko.c	2015-04-27 10:11:18.684249096 +0200
@@ -105,6 +105,8 @@
      If this assumption is incorrect, please report the bug to
      bug-gnulib.  */
   if (0)
+#elif defined _PDCLIB_C_VERSION     /* PDCLIB */
+  if (fp->bufidx == 0)
 #else
   #error "Please port gnulib fseeko.c to your platform! Look at the code in fseeko.c, then report this to bug-gnulib."
 #endif
diff -rubB coreutils-8.23.org/lib/fseterr.c coreutils-8.23/lib/fseterr.c
--- coreutils-8.23.org/lib/fseterr.c	2014-05-29 14:05:50.000000000 +0200
+++ coreutils-8.23/lib/fseterr.c	2015-04-27 10:11:33.612249553 +0200
@@ -48,6 +48,8 @@
 #elif defined EPLAN9                /* Plan9 */
   if (fp->state != 0 /* CLOSED */)
     fp->state = 5 /* ERR */;
+#elif defined _PDCLIB_C_VERSION     /* PDCLIBC */
+  fp->status |= _PDCLIB_ERRORFLAG;
 #elif 0                             /* unknown  */
   /* Portable fallback, based on an idea by Rich Felker.
      Wow! 6 system calls for something that is just a bit operation!
diff -rubB coreutils-8.23.org/lib/freadseek.c coreutils-8.23/lib/freadseek.c
--- coreutils-8.23.org/lib/freadseek.c	2014-05-29 14:05:50.000000000 +0200
+++ coreutils-8.23/lib/freadseek.c	2015-04-27 10:45:59.532312816 +0200
@@ -62,6 +62,10 @@
   fp->__bufp += increment;
 #elif defined EPLAN9                /* Plan9 */
   fp->rp += increment;
+#elif defined _PDCLIB_C_VERSION     /* PDCLIB */
+  while (increment > 0 && fp->ungetidx > 0)
+    fp->ungetidx--, increment--;
+  fp->bufidx += increment;
 #elif defined SLOW_BUT_NO_HACKS     /* users can define this */
 #else
  #error "Please port gnulib freadseek.c to your platform! Look at the definition of getc, getc_unlocked on your system, then report this to bug-gnulib."
diff -rubB coreutils-8.23.org/lib/freadahead.c coreutils-8.23/lib/freadahead.c
--- coreutils-8.23.org/lib/freadahead.c	2014-05-29 14:05:50.000000000 +0200
+++ coreutils-8.23/lib/freadahead.c	2015-04-27 10:51:32.896323024 +0200
@@ -84,6 +84,10 @@
   if (fp->state == 4 /* WR */ || fp->rp >= fp->wp)
     return 0;
   return fp->wp - fp->rp;
+#elif defined _PDCLIB_C_VERSION     /* PDCLIBC */
+  if ((fp->status & _PDCLIB_FREAD) == 0)
+    return 0;
+  return fp->bufend - fp->bufidx;
 #elif defined SLOW_BUT_NO_HACKS     /* users can define this */
   abort ();
   return 0;
diff -rubB coreutils-8.23.org/lib/freadptr.c coreutils-8.23/lib/freadptr.c
--- coreutils-8.23.org/lib/freadptr.c	2014-05-29 14:05:50.000000000 +0200
+++ coreutils-8.23/lib/freadptr.c	2015-04-27 10:54:44.192328882 +0200
@@ -108,6 +108,11 @@
     return NULL;
   *sizep = fp->wp - fp->rp;
   return fp->rp;
+#elif defined _PDCLIB_C_VERSION     /* PDCLIBC */
+  if ((fp->status & _PDCLIB_FREAD) == 0)
+    return 0;
+  *sizep = fp->bufend - fp->bufidx;
+	return fp->bufidx;
 #elif defined SLOW_BUT_NO_HACKS     /* users can define this */
   /* This implementation is correct on any ANSI C platform.  It is just
      awfully slow.  */
diff -rubB coreutils-8.23.org/configure.ac coreutils-8.23/configure.ac
--- coreutils-8.23.org/configure.ac	2014-07-18 03:40:57.000000000 +0200
+++ coreutils-8.23/configure.ac	2015-04-27 11:13:42.884363751 +0200
@@ -464,6 +464,7 @@
     void __attribute__ ((constructor))
     stdbuf_init (void)
     {
+#error Unsupported (needs dynamic library support)
       #if __ELF__
         stdbuf = 1;
       #endif
diff -rubB coreutils-8.23.org/Makefile.am coreutils-8.23/Makefile.am
--- coreutils-8.23.org/Makefile.am	2014-07-14 00:09:52.000000000 +0200
+++ coreutils-8.23/Makefile.am	2015-04-27 11:15:38.556367293 +0200
@@ -214,5 +214,4 @@
 include $(top_srcdir)/lib/local.mk
 include $(top_srcdir)/src/local.mk
 include $(top_srcdir)/doc/local.mk
-include $(top_srcdir)/man/local.mk
 include $(top_srcdir)/tests/local.mk
