/* { dg-do preprocess } */
/* { dg-options "-std=gnu++98 -pedantic" } */

ª
« /* { dg-error "not valid in an identifier" } */
¶ /* { dg-error "not valid in an identifier" } */
º
À
Ö
΄ /* { dg-error "not valid in an identifier" } */

٩ /* { dg-error "not valid at the start of an identifier" } */
A٩
0º
0٩
๙ /* { dg-error "not valid at the start of an identifier" } */
A๙
