diff -rubB make-3.82.patched/arscan.c make-3.82/arscan.c
--- make-3.82.patched/arscan.c	2010-07-13 03:20:32.000000000 +0200
+++ make-3.82/arscan.c	2011-03-05 19:41:25.464855000 +0100
@@ -252,8 +252,9 @@
 # define __AR_BIG__
 #endif
 
-#ifndef WINDOWS32
-# ifndef __BEOS__
+/* XXX Ananas doesn't have ar.h, so we'll just use the BeOS one instead */
+#if !defined(WINDOWS32) || defined(__Ananas__)
+# if !defined(__BEOS__) && !defined(__Ananas__)
 #  include <ar.h>
 # else
    /* BeOS 5 doesn't have <ar.h> but has archives in the same format
diff -rubB make-3.82.patched/config/config.sub make-3.82/config/config.sub
--- make-3.82.patched/config/config.sub	2010-07-28 07:42:11.000000000 +0200
+++ make-3.82/config/config.sub	2011-03-05 19:02:05.904854997 +0100
@@ -1452,6 +1452,8 @@
 		;;
         -nacl*)
 	        ;;
+	-ananas*)
+		;;
 	-none)
 		;;
 	*)
diff -rubB make-3.82.patched/main.c make-3.82/main.c
--- make-3.82.patched/main.c	2010-07-19 09:10:53.000000000 +0200
+++ make-3.82/main.c	2011-03-05 19:19:42.994854993 +0100
@@ -3083,7 +3083,7 @@
   time_t when;
 
   when = time ((time_t *) 0);
-  printf (_("\n# Make data base, printed on %s"), ctime (&when));
+  printf (_("\n# Make data base, printed on %s"), "XXXTODO" /* ctime (&when) */);
 
   print_variable_data_base ();
   print_dir_data_base ();
@@ -3093,7 +3093,7 @@
   strcache_print_stats ("#");
 
   when = time ((time_t *) 0);
-  printf (_("\n# Finished Make data base on %s\n"), ctime (&when));
+  printf (_("\n# Finished Make data base on %s\n"), "XXXTODO" /* ctime (&when) */);
 }
 
 static void
